include/pmacc/lockstep/ForEach.hpp

```diff
@@ -0,0 +1,287 @@
+        class ForEach<Config<T_domainSize, T_numWorkers, T_simdSize>>
+            : Config<T_domainSize, T_numWorkers, T_simdSize>
+            , Worker<T_numWorkers>
+        {
+            /* Helper to check if a member exists
+             *
+             * Derived from C++17 std::void_t.
+             * This implementation will be removed with Void provided by alpaka 0.6.0 release (not included in the
+             * 0.6.0rc3 we currently using).
+             */
+            template<class...>
+            using Void = void;
+
+            /** Get the result of a functor invocation.
+             *
+             * @attention The behavior is undefined for ill-formed invocations.
+             *
+             * @{
+             */
+#if BOOST_COMP_CLANG_CUDA && __CUDACC_VER_MAJOR__ <= 10
+            template<typename F, typename... T_Args>
+            using InvokeResult_t = decltype(alpaka::core::declval<F>()(alpaka::core::declval<T_Args>()...));
+#else
+#    if __cplusplus >= 201703L
+            template<typename F, typename... T_Args>
+            using InvokeResult_t = typename std::invoke_result<F, T_Args...>::type;
+#    else
+            template<typename F, typename... T_Args>
+            using InvokeResult_t = typename std::result_of<F(T_Args...)>::type;
+#    endif
+#endif
+            /**@}*/
+
+        public:
+            using BaseConfig = Config<T_domainSize, T_numWorkers, T_simdSize>;
+
+            using BaseConfig::domainSize;
+            using BaseConfig::numCollIter;
+            using BaseConfig::numWorkers;
+            using BaseConfig::simdSize;
+
+            /** constructor
+             *
+             * @param workerIdx index of the worker: range [0;workerSize)
+             */
+            HDINLINE
+            ForEach(uint32_t const workerIdx) : Worker<numWorkers>(std::move(workerIdx))
+            {
+            }
+
+            HDINLINE Worker<numWorkers> getWorkerCfg() const
+            {
+                return static_cast<Worker<numWorkers>>(*this);
+            }
+
+            /** execute a functor
+             *
+             * Distribute the indices of the domain even over all worker and execute a user defined functor.
+             * There is no guarantee in which order the indices will be processed.
+             *
+             * @param functor is called for each index which is mapped to the worker
+             * @{
+             */
+
+            /** The functor must fulfill the following interface:
+             * @code
+             * void operator()( lockstep::Idx< T_domainSize > const idx );
+             * // or
+             * void operator()( uint32_t const linearIdx );
+             * @endcode
+             */
+            template<
+                typename T_Functor,
+                // check if functor is invocable
+                typename = Void<decltype(alpaka::core::declval<T_Functor>()(alpaka::core::declval<Idx const>()))>,
```

Comment:
I think it would be a little easier to read if `decltype(alpaka::core::declval<T_Functor>()(alpaka::core::declval<Idx const>()))` or maybe even better a `Void<>` of that became its own named alias like
```cpp
template<typename T_Functor>
using IsCallableWithIndex = Void<decltype(alpaka::core::declval<T_Functor>()(alpaka::core::declval<Idx const>()))>;
```

But in terms of code lines that would only increase it, so do as you like.

Reply:
Let us merge this PR as it is because I know  this is working on different compiler.
I will refactor the foreach in a follow up anyway and will test your suggestion too.

CHanging void is not necessary because it will be removed with alapaks `Void` as soon as we have alpaka 0.6.0 or 0.6.1 in PIConGPU.

Reply:
I marked it unresolved to find it faster, else I need to check all resolved comment boxes